canonical-web-and-design/maas-ui
# Add betterer and strict compilation test.

[Betterer](https://github.com/phenomnomnominal/betterer) is a tool for snapshotting any given metric to ensure that metric either gets better(er) as new code lands. For example, it could be used to monitor the performance of a particular component, or count the number of database queries. 

In this branch, betterer allows us to run a test with a modified `.tsconfig` with `strict` enabled. It then snapshots the results of that test run. This allows us to track our progress towards `strict` mode, and ensures we don't move backwards from that goal. 

As this is a new tool/idea, feel free to discuss if we want to give this a go in the comments.

## QA
As we don't have any TypeScript files with issues currently, you'll have to artificially create one to see how betterer works. Rename a .js file to .ts and run `yarn betterer` (we don't have `checkJs` enabled, only TypeScript files get checked by the compiler).

If the file has warnings in `strict` mode, you should see them appear in `.betterer.results`. If the number of warnings decreases, betterer will report that our progress towards the `strict` mode goal is improving, otherwise it will be treated as a lint failure.

### Example
The first time you run betterer, you should see:

![image](https://user-images.githubusercontent.com/130286/84730261-caa0d480-afe9-11ea-9eb8-263ae375037e.png)

Running betterer again will inform you that nothing changed:

![image](https://user-images.githubusercontent.com/130286/84730295-e6a47600-afe9-11ea-8372-b9c9badf4dcf.png)

After fixing an issue, betterer will inform you that 1 test improved:

![image](https://user-images.githubusercontent.com/130286/84730314-f58b2880-afe9-11ea-881d-52e01e4ffd1a.png)

And finally, when all the issues are fixed, we'll have met our goal:

![image](https://user-images.githubusercontent.com/130286/84730399-2bc8a800-afea-11ea-99fc-2c36929f05db.png)
